crates/ra_parser/src/grammar/patterns.rs

```diff
@@ -274,3 +269,29 @@ fn bind_pat(p: &mut Parser, with_at: bool) -> CompletedMarker {
     }
     m.complete(p, BIND_PAT)
 }
+
+// test_err ref_box_pat
+// fn main() {
+//     let ref box i = ();
+// }
+
+// test box_pat
+// struct Outer<'a> {
+//     i: Box<i32>,
+//     j: Box<Inner<'a>>,
+// }
+//
+// struct Inner<'a>(&'a i32);
```

Comment:
Let's remove `struct Ounter` and `struct Inner`, they are completely irrelvent for testing syntax

Reply:
So it seems that this test only passes if there's something named `Inner` and `Outer` in scope. This is probably a shortcoming of the parsing code?

Reply:
It shouldn't be the case. Judging by the actual .rs file in the `inline` folder, looks like the test harness is confused over a blank line?

Reply:
I get the following error if the names aren't in scope:
```
thread 'tests::parser_tests' panicked at 'assertion failed: `(left == right)`
  left: `[SyntaxError { kind: ParseError(ParseError("expected a name")), location: 52 }, SyntaxError { kind: ParseError(ParseError("expected COMMA")), location: 55 }]`,
 right: `[]`: There should be no errors in the file "/home/mackendy/src/rust/rust-analyzer/crates/ra_syntax/test_data/parser/inline/ok/0143_box_pat.rs"', crates/ra_syntax/src/tests.rs:23:9
```

0143_box_pat.rs:

```rust
fn main() {
    let box i = ();
    let box Outer { box i, j: box Inner(box &x) } = ();
    let box ref mut i = ();
}
```

Reply:
I needed to extend `record_field_pat_list` to handle the `box i` part. This syntax is equivalent to `i: box i`. I was actually a little surprised that [it worked](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e5b6b8b56f190602c9e19c8bc1be5ee0).